### PR TITLE
Support validation of enrollment.

### PR DESCRIPTION
**Pull request by brianhw**
Includes:
- Synthesize validation events from Sqoop dumps of student_courseenrollment.
- Compare validation events with enroll/unenroll events.   Includes unit tests.
  Supports additional flags for:
  - Outputting tuples (for debugging) or synthetic events (for production)
  - Ignoring any events before the first validation event (for each user/course pair).  

The latter permits checking for errors in an incremental manner, by using an interval to specify a tighter period of time (e.g. a day or a few days) which includes a set of enrollment validation events.   The run can then test the consistency of events since the enrollment validation, up to the end of the interval (which selects all events included).  This could be set up to run on a daily basis, to help flag inconsistencies more quickly as well as to generate new synthetic events in an incremental manner.

### CODE REVIEW

**Comment by brianhw** [comment]
@mulby Ready for initial review of the approach.

File: edx/analytics/tasks/enrollment_validation.py [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+from edx.analytics.tasks.util.datetime_util import add_microseconds, mysql_datetime_to_isoformat
+from edx.analytics.tasks.util.event_factory import SyntheticEventFactory
+from edx.analytics.tasks.util.hive import WarehouseMixin
+
+
+log = logging.getLogger(__name__)
+
+DEACTIVATED = 'edx.course.enrollment.deactivated'
+ACTIVATED = 'edx.course.enrollment.activated'
+VALIDATED = 'edx.course.enrollment.validated'
```
Comment (mulby) [inline edx/analytics/tasks/enrollment_validation.py:28]: As was mentioned somewhere else, I think we should try to find a different name for this.

File: edx/analytics/tasks/enrollment_validation.py [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+SENTINEL = 'edx.course.enrollment.sentinel'
```
Comment (mulby) [inline edx/analytics/tasks/enrollment_validation.py]: Also, a different name for this.

File: edx/analytics/tasks/enrollment_validation.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+        if course_id is None or not opaque_key_util.is_valid_course_id(course_id):
+            log.error("encountered explicit enrollment event with invalid course_id: %s", event)
+            return
+
+        user_id = event_data.get('user_id')
+        if user_id is None:
+            log.error("encountered explicit enrollment event with no user_id: %s", event)
+            return
+
+        mode = event_data.get('mode')
```
Comment (mulby) [inline edx/analytics/tasks/enrollment_validation.py:126]: If we aren't handling mode yet, do we actually need to extract this?

File: edx/analytics/tasks/enrollment_validation.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+        self.lower_bound_date_string = kwargs.get('lower_bound_date_string')
+
+        if self.event_output:
+            self.factory = SyntheticEventFactory(
+                event_source='server',
+                synthesizer='enrollment_validation',
+            )
+            self.generate_output = self.synthetic_event
+        else:
+            self.generate_output = self.create_tuple
```
Comment (mulby) [inline edx/analytics/tasks/enrollment_validation.py]: I wonder if we couldn't simplify this by removing the tuple mode. I have no problem curating raw JSON events...

File: edx/analytics/tasks/enrollment_validation.py [hunk 5]
```diff
@@ -0,0 +1,817 @@
...
+
+log = logging.getLogger(__name__)
+
+# Product event_type values:
+DEACTIVATED = 'edx.course.enrollment.deactivated'
+ACTIVATED = 'edx.course.enrollment.activated'
+MODE_CHANGED = 'edx.course.enrollment.mode_changed'
+
+# Validation-event event_type values:
+VALIDATED = 'edx.course.enrollment.validated'
```
Comment (mulby) [inline edx/analytics/tasks/enrollment_validation.py:28]: edx.course.enrollment.snapshot?

**Comment by mulby** [comment]
I don't see any blockers here, OK with leaving the "validated" event

:+1: once we find and fix the bug

File: edx/analytics/tasks/enrollment_validation.py (hunk 3 above)
Comment (brianhw) [inline edx/analytics/tasks/enrollment_validation.py:126]: Actually, we are handling mode yet.  Originally to make sure that the right mode appeared on the synthetic event being output.  But it's also outputting synthetic mode-change events as well.

File: edx/analytics/tasks/enrollment_validation.py (hunk 2 above)
Comment (brianhw) [inline edx/analytics/tasks/enrollment_validation.py]: Done.

File: edx/analytics/tasks/enrollment_validation.py (hunk 5 above)
Comment (brianhw) [inline edx/analytics/tasks/enrollment_validation.py:28]: I like that.  I'll switch it when the validation work is done, and there's a reasonable point to recalculate the validation events.

**Pull request merged**